sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -345,6 +348,28 @@ private[sql] class ParquetRelation2(
     // Schema of the whole table, including partition columns.
     var schema: StructType = _
 
+    def filterDataStatusesWithoutSummaries(
+      leaves: Seq[FileStatus],
+      dataStatuses: Seq[FileStatus]): Seq[FileStatus] = {
+
+      // Get the partitions that have summary files
+      val typeInference = sqlContext.conf.partitionColumnTypeInferenceEnabled()
+      val summariesPaths = metadataStatuses.map(_.getPath.getParent()) ++
+        commonMetadataStatuses.map(_.getPath.getParent())
+      val summariesPartitions = PartitioningUtils.parsePartitions(summariesPaths,
+        PartitioningUtils.DEFAULT_PARTITION_NAME, typeInference).partitions.toSet
+
+      dataStatuses.filterNot { d =>
+        val part = PartitioningUtils.parsePartitions(Seq(d.getPath.getParent()),
+          PartitioningUtils.DEFAULT_PARTITION_NAME, typeInference)
+        if (summariesPartitions.size > 0 && part.partitions.length > 0) {
+          summariesPartitions.contains(part.partitions(0))
+        } else {
+          false
+        }
+      }
+    }
```

Comment:
Under current implementation, the summary files and part-files are all scanned and merged. I think it is accepted and can guarantee to deal possible problems.